File: src/frontend/src/types/Plant.ts
import type { PlantType } from './PlantType';

/** Localised display fields (common name + short description) per language. */
export interface PlantTranslation {
  id: number;
  language: string;
  commonName: string;
  description: string | null;
}

/** A categorised photo with licensing/attribution and per-source ordering. */
export interface PlantImage {
  id: number;
  imageType: string;
  url: string;
  thumbnailUrl: string | null;
  width: number | null;
  height: number | null;
  licenseName: string | null;
  licenseUrl: string | null;
  credit: string | null;
  source: string;
  sourceExternalId: string | null;
  displayOrder: number;
  isFlagged: boolean;
  /**
   * Server-composed, always-non-null attribution line shipped by the detail DTO
   * via `ImageAttribution.Compose` (format "© credit — license"). NOT rendered
   * verbatim by the gallery: its format differs from the approved design
   * ("© credit · source · license"), so `PlantGallerySection` composes its own
   * line from credit/source/licenseName. Aligning the backend format is SMA-180.
   */
  attribution: string;
}

/** Long-form rich description in a single language, one row per locale. */
export interface PlantLongDescription {
  id: number;
  language: string;
  longDescription: string;
  sourceMethod: string | null;
}

/** A vernacular name in one language, with `isPrimary` flagging the preferred entry. */
export interface PlantCommonName {
  id: number;
  languageCode: string;
  name: string;
  isPrimary: boolean;
}

/** A pest or pathogen affecting the plant, sourced from Perenual today. */
export interface PlantPest {
  id: number;
  name: string;
  type: string;
  description: string | null;
  symptoms: string | null;
  solutions: string | null;
  imageUrl: string | null;
  source: string;
  sourceExternalId: string | null;
}

/** A historical / alternative scientific name used during ETL fuzzy match. */
export interface PlantSynonym {
  id: number;
  synonym: string;
  authority: string | null;
}

/** Cross-reference to an external taxonomy/enrichment API, with link metadata. */
export interface PlantSource {
  id: number;
  sourceType: string;
  externalId: string;
  url: string | null;
  notes: string | null;
  lastFetchedAt: string | null;
}

/** Trefle-specific structured data (1-1 with Plant); `RawResponseJson` is omitted from the DTO. */
export interface PlantTrefleData {
  id: string;
  trefleSlug: string | null;
  wfoId: string | null;
  growthHabit: string | null;
  flowerColors: string | null;
  foliageColors: string | null;
  nativeRegionsJson: string | null;
  introducedRegionsJson: string | null;
  soilNutrimentsLevel: number | null;
  soilSalinityLevel: number | null;
  atmosphericHumidityLevel: number | null;
  apiVersion: string | null;
  lastSyncAt: string;
}

/** Perenual-specific structured data (1-1 with Plant); `RawResponseJson` is omitted from the DTO. */
export interface PlantPerenualData {
  id: string;
  perenualId: number;
  /**
   * Id originally requested from Perenual. Differs from `perenualId` (the
   * canonical id returned by `response.id`) when Perenual rewrites the
   * server-side id mapping (cf. backend issue #67). Frontend prefers this
   * value when building user-facing public URLs so the link lands on the
   * correct species page. `null` on rows enriched before this column existed
   * — caller falls back to `perenualId`.
   */
  requestedPerenualId: number | null;
  cultivar: string | null;
  perenualType: string | null;
  originCountries: string | null;
  propagationMethods: string | null;
  wateringBenchmark: string | null;
  wateringBenchmarkUnit: string | null;
  sunlightPreferences: string | null;
  pruningMonths: string | null;
  maintenance: string | null;
  floweringSeason: string | null;
  harvestSeason: string | null;
  hasEdibleFruit: boolean | null;
  hasEdibleLeaves: boolean | null;
  isCulinary: boolean | null;
  plantAnatomyJson: string | null;
  apiVersion: string | null;
  hasSupremeData: boolean;
  lastSyncAt: string;
  // Perenual Supreme xData (Sprint 1.5 PR B) — surfaced in Section F.6 (Phase 3).
  xWateringBasedTempMinC: number | null;
  xWateringBasedTempMaxC: number | null;
  xWateringPhMin: number | null;
  xWateringPhMax: number | null;
  xSunlightHoursMin: number | null;
  xSunlightHoursMax: number | null;
  xTemperatureToleranceMinC: number | null;
  xTemperatureToleranceMaxC: number | null;
  xPlantSpacingValue: number | null;
  xPlantSpacingUnit: string | null;
  xWateringQualityJson: string | null;
  xWateringPeriodJson: string | null;
}

/**
 * Mirrors `SmartCrops.Api.Dtos.PlantDetailResponse`. Enum-typed scalars arrive
 * as strings (the backend mapper applies `.ToString()`); the frontend resolves
 * human labels via `plantDetail.enumValues.*` i18n keys.
 */
export interface Plant {
  id: string;
  scientificName: string;
  plantTypeId: number;
  plantType: PlantType | null;

  sunExposure: string | null;
  waterNeeds: string | null;
  sowingPeriod: string | null;
  harvestPeriod: string | null;
  imageUrl: string | null;
  /** Attribution line for {@link imageUrl} — only the list DTO carries it (the detail view attributes per gallery image), so optional. */
  imageAttribution?: string | null;
  /** Localised common name (list DTO, `?lang=`); optional — detail uses {@link translations}/{@link commonNames} instead. */
  commonName?: string | null;
  /** Localised short description (list DTO, `?lang=`); optional. */
  description?: string | null;

  gbifTaxonKey: number | null;
  family: string | null;
  genus: string | null;
  speciesEpithet: string | null;
  author: string | null;
  wfoId: string | null;
  year: number | null;

  lifeCycle: string | null;
  growthRate: string | null;
  wateringNeedLevel: string | null;
  careLevel: string | null;
  growthHabit: string | null;

  hardinessZoneMin: number | null;
  hardinessZoneMax: number | null;
  minHeightCm: number | null;
  maxHeightCm: number | null;
  minSpreadCm: number | null;
  maxSpreadCm: number | null;
  soilPhMin: number | null;
  soilPhMax: number | null;
  lightLevel: number | null;
  soilNutriments: number | null;
  minTempC: number | null;
  maxTempC: number | null;

  isEdible: boolean | null;
  isVegetable: boolean | null;
  isMedicinal: boolean | null;
  isIndoor: boolean | null;
  isDroughtTolerant: boolean | null;
  isSaltTolerant: boolean | null;
  isThorny: boolean | null;
  isInvasive: boolean | null;
  isTropical: boolean | null;
  isToxicToHumans: boolean | null;
  isToxicToPets: boolean | null;
  attractsPollinators: boolean | null;

  flowerColors: string | null;
  nativeRegions: string | null;
  introducedRegions: string | null;
  edibleParts: string | null;
  sowingInstructions: string | null;
  propagationInstructions: string | null;

  enrichmentSources: readonly string[];
  lastEnrichmentAt: string | null;

  createdAt: string;
  updatedAt: string;

  // Collections mirror the backend `IReadOnlyList<T>` contract — surfacing
  // them as `readonly` here prevents callers from accidentally mutating
  // shared Plant state (e.g. via `plant.images.sort(...)`); use a spread
  // (`[...plant.images].sort()`) or array methods that return a new array.
  // Optional: the neutral list DTO (PlantListItemResponse, PR #100) omits
  // translations, so list-sourced Plant objects carry none at runtime. Marked
  // optional so the type forces a guard at every access site (SMA-73).
  translations?: readonly PlantTranslation[];
  images: readonly PlantImage[];
  longDescriptions: readonly PlantLongDescription[];
  commonNames: readonly PlantCommonName[];
  pests: readonly PlantPest[];
  synonyms: readonly PlantSynonym[];
  sources: readonly PlantSource[];

  trefleData: PlantTrefleData | null;
  perenualData: PlantPerenualData | null;
}

File: src/backend/SmartCrops.Core/Entities/Plant.cs
using SmartCrops.Core.Enums;
using SmartCrops.Core.Interfaces;

namespace SmartCrops.Core.Entities;

/// <summary>
/// Root of the plant aggregate.
///
/// Architecture:
/// - Canonical fields (life cycle, watering need, hardiness, etc.) live directly on
///   Plant for fast reads — denormalized READ MODEL.
/// - Source-specific raw data is kept in <see cref="PlantTrefleData"/> and
///   <see cref="PlantPerenualData"/> as an audit trail and for re-derivation.
/// - Multilingual content lives in <see cref="PlantTranslation"/> (legacy short text)
///   and <see cref="PlantLongDescription"/> (rich long-form, one row per language).
/// - During ETL, merge priority is: Manual &gt; Perenual &gt; Trefle &gt; GBIF.
/// </summary>
public class Plant : IHasUpdatedAt
{
    public Guid Id { get; set; }

    /// <summary>
    /// Latin scientific name — language-neutral, never translated (e.g. "Solanum lycopersicum").
    /// </summary>
    public required string ScientificName { get; set; }

    public int PlantTypeId { get; set; }
    public PlantType PlantType { get; set; } = null!;

    /// <summary>
    /// Enum-like key resolved by the UI (e.g. "full_sun", "partial_shade", "full_shade").
    /// Storing a key instead of an enum keeps the DB schema stable when values are added.
    /// </summary>
    public string? SunExposure { get; set; }

    /// <summary>
    /// Enum-like key resolved by the UI (e.g. "low", "moderate", "regular", "high").
    /// </summary>
    public string? WaterNeeds { get; set; }

    /// <summary>
    /// Human-readable period key (e.g. "march-may"). Kept as a string so it can express
    /// ranges, hemisphere variants, or climate-zone qualifiers without schema changes.
    /// </summary>
    public string? SowingPeriod { get; set; }

    /// <summary>
    /// Human-readable period key (e.g. "july-october").
    /// </summary>
    public string? HarvestPeriod { get; set; }

    public string? ImageUrl { get; set; }

    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }

    // ── Identity (GBIF canonical) ───────────────────────────────────────────────

    /// <summary>GBIF taxon key — globally unique numeric ID for the species.</summary>
    public int? GbifTaxonKey { get; set; }

    /// <summary>Family name (e.g. "Solanaceae").</summary>
    public string? Family { get; set; }

    /// <summary>Genus name (e.g. "Solanum").</summary>
    public string? Genus { get; set; }

    /// <summary>Species epithet (e.g. "lycopersicum").</summary>
    public string? SpeciesEpithet { get; set; }

    /// <summary>Taxonomic authority (e.g. "L." for Linnaeus).</summary>
    public string? Author { get; set; }

    /// <summary>
    /// World Flora Online taxon id (audit source for the canonical species).
    /// Denormalized from <see cref="PlantTrefleData"/> per ADR-0003 so the
    /// canonical read model carries every cross-reference id directly.
    /// </summary>
    public string? WfoId { get; set; }

    /// <summary>
    /// Year the species was first published in scientific literature (GBIF).
    /// CHECK constraint: between 1700 and the current year.
    /// </summary>
    public int? Year { get; set; }

    /// <summary>
    /// Rank at which this plant's identity is pinned. Defaults to
    /// <see cref="PlantTaxonRank.Species"/>; set to <see cref="PlantTaxonRank.Genus"/>
    /// when only a genus-level identity is defensible (horticultural group or
    /// trade designation with no resolvable accepted species). Such rows usually
    /// also carry <see cref="IdentityNeedsReview"/>.
    /// </summary>
    public PlantTaxonRank TaxonRank { get; set; } = PlantTaxonRank.Species;

    // ── Canonical READ MODEL (denormalized from sources) ────────────────────────

    public PlantLifeCycle? LifeCycle { get; set; }
    public PlantGrowthRate? GrowthRate { get; set; }
    public PlantWateringNeed? WateringNeedLevel { get; set; }
    public PlantCareLevel? CareLevel { get; set; }

    /// <summary>USDA hardiness zone lower bound.</summary>
    public int? HardinessZoneMin { get; set; }

    /// <summary>USDA hardiness zone upper bound.</summary>
    public int? HardinessZoneMax { get; set; }

    public int? MinHeightCm { get; set; }
    public int? MaxHeightCm { get; set; }
    public int? MinSpreadCm { get; set; }
    public int? MaxSpreadCm { get; set; }

    public decimal? SoilPhMin { get; set; }
    public decimal? SoilPhMax { get; set; }

    /// <summary>Light requirement on a 0-10 scale (Trefle convention).</summary>
    public int? LightLevel { get; set; }

    /// <summary>Trefle soil nutriments preference on a 0-10 scale.</summary>
    public int? SoilNutriments { get; set; }

    public int? MinTempC { get; set; }
    public int? MaxTempC { get; set; }

    // ── Boolean flags ───────────────────────────────────────────────────────────

    public bool? IsEdible { get; set; }

    /// <summary>Trefle classification: whether the plant is a vegetable.</summary>
    public bool? IsVegetable { get; set; }

    public bool? IsMedicinal { get; set; }
    public bool? IsIndoor { get; set; }
    public bool? IsDroughtTolerant { get; set; }
    public bool? IsSaltTolerant { get; set; }
    public bool? IsThorny { get; set; }
    public bool? IsInvasive { get; set; }
    public bool? IsTropical { get; set; }
    public bool? IsToxicToHumans { get; set; }
    public bool? IsToxicToPets { get; set; }
    public bool? AttractsPollinators { get; set; }

    // ── Growth habit (Trefle) ───────────────────────────────────────────────────

    /// <summary>
    /// Plant growth habit per Trefle classification. Stored as string in DB.
    /// </summary>
    public PlantGrowthHabit? GrowthHabit { get; set; }

    // ── Trefle structured data (denormalized from PlantTrefleData per ADR-0003) ─

    /// <summary>
    /// Flower colors as JSON array of strings (e.g. <c>["red","pink"]</c>).
    /// Denormalized from <see cref="PlantTrefleData.FlowerColors"/> for fast
    /// Library filtering without cross-joining the audit table.
    /// </summary>
    public string? FlowerColors { get; set; }

    /// <summary>
    /// TDWG region codes where the plant is native, as JSON array.
    /// Denormalized from <see cref="PlantTrefleData.NativeRegionsJson"/>.
    /// </summary>
    public string? NativeRegions { get; set; }

    /// <summary>
    /// TDWG region codes where the plant has been introduced, as JSON array.
    /// Denormalized from <see cref="PlantTrefleData.IntroducedRegionsJson"/>.
    /// </summary>
    public string? IntroducedRegions { get; set; }

    // ── Perenual descriptive data ───────────────────────────────────────────────

    /// <summary>
    /// Edible parts as JSON array of strings (e.g. <c>["fruit","leaf","root"]</c>).
    /// </summary>
    public string? EdibleParts { get; set; }

    /// <summary>Sowing instructions text (Perenual).</summary>
    public string? SowingInstructions { get; set; }

    /// <summary>Propagation instructions text (Perenual).</summary>
    public string? PropagationInstructions { get; set; }

    // ── Enrichment metadata ─────────────────────────────────────────────────────

    public EnrichmentStatus EnrichmentStatus { get; set; } = EnrichmentStatus.Manual;

    public DateTime? LastEnrichmentAt { get; set; }

    /// <summary>
    /// Flags a plant whose taxonomic identity is provisional and needs an admin
    /// to confirm or correct it — e.g. pinned at <see cref="PlantTaxonRank.Genus"/>,
    /// or resolved from an ambiguous source name. Defaults to <c>false</c>.
    /// </summary>
    public bool IdentityNeedsReview { get; set; }

    /// <summary>
    /// Perenual id this plant was originally requested under, denormalised from
    /// <see cref="PlantPerenualData.RequestedPerenualId"/> for fast query without
    /// joining the audit table. Differs from <see cref="PlantPerenualData.PerenualId"/>
    /// when Perenual's server-side canonicalisation rewrites the id (cf. issue #67:
    /// requesting <c>/details/8759</c> returns <c>response.id = 8758</c>, which
    /// happens to be a wholly different species' record). User-facing public
    /// URLs use this value instead of the canonical id so the link lands on the
    /// correct species page. Rows enriched before this column existed stay
    /// <c>null</c>; the frontend falls back to the canonical id in that case.
    /// </summary>
    public int? RequestedPerenualId { get; set; }

    // ── Navigation properties ───────────────────────────────────────────────────

    public ICollection<PlantTranslation> Translations { get; set; } = [];
    public ICollection<PlantSuggestion> Suggestions { get; set; } = [];

    public ICollection<PlantLongDescription> LongDescriptions { get; set; } = [];
    public ICollection<PlantCommonName> CommonNames { get; set; } = [];
    public ICollection<PlantImage> Images { get; set; } = [];
    public ICollection<PlantPhase> Phases { get; set; } = [];
    public ICollection<PlantSynonym> Synonyms { get; set; } = [];
    public ICollection<PlantSource> Sources { get; set; } = [];
    public ICollection<PlantPest> Pests { get; set; } = [];

    public PlantTrefleData? TrefleData { get; set; }
    public PlantPerenualData? PerenualData { get; set; }
}
